# Python Numpy

### Numerical Python and is the core library for numeric and scientific computing
The most basic object in NumPy is the ```array```, which is **homogenous** in nature. By homogenous, we mean that all the elements in a numpy array have to be of the **same data type**, which is commonly numeric (float or integer). 

In [3]:
# Import library
import numpy as np

### Create numpy array

In [4]:
a = np.array([1, 2, 3])
b = np.array([[1, 2, 3], [4, 5, 6]])
print(a)
print(b)

[1 2 3]
[[1 2 3]
 [4 5 6]]


In [5]:
# Get dimension
print(a.ndim)       # Result is 1 - 1D array
print(b.ndim)       # Result is 2 - 2D array

1
2


In [6]:
# Get shape
print(a.shape)
print(b.shape)

(3,)
(2, 3)


In [7]:
# Get type of array

print(a.dtype)
print(b.dtype)

int32
int32


In [8]:
# Difference between size and itemsize
# size : int - Number of elements in the array. 
# itemsize : int - The memory use of each array element in bytes

print("For array 'a': Contains total", a.size, "Element and", a.itemsize, "byte")
print("For array 'b': Contains total", b.size, "Element and", b.itemsize, "byte")

For array 'a': Contains total 3 Element and 4 byte
For array 'b': Contains total 6 Element and 4 byte


### Accessing or changing specific element in numpy array

In [9]:
a = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [10]:
# Get the specific element, e.g- a[row_index, column_index]

a[1, 2]

7

In [11]:
# Get the specific row

a[1, :]

array([5, 6, 7, 8])

In [12]:
# Get the specific column

a[:, 2]

array([ 3,  7, 11])

In [13]:
# Accessing with step-up value
# [row_index, start_index:end_index:stepup_value]

a[1, 0:-1:2]

array([5, 7])

## Creating Numpy Array


 - ```np.ones()```: Create an array of 1s.
 - ```np.zeros()```: Create an array of 0s.
 - ```np.arange()```: Create an array with increments of fixed step size.
 - ```np.linspace()```: Create an array of fixed length.

In [107]:
# creating an array with 4 elements and increments of a fixed step size
# default step size is one
np_1 = np.arange(1, 5)
print(np_1)

# creating an array of zeroes with 5 elements
np_2 = np.zeros(5)
np_3 = np.zeros(4, dtype="int")
print(np_2)
print(np_3)

[1 2 3 4]
[0. 0. 0. 0. 0.]
[0 0 0 0]


In [14]:
# All 0s matrix

np.zeros((2, 3))         # 2 rows and 3 columns

array([[0., 0., 0.],
       [0., 0., 0.]])

In [15]:
# All 1s matrix

np.ones((2, 3, 3))

array([[[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]],

       [[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]]])

In [110]:
# creating an array of fixed length
np_4 = np.linspace(1, 10, 9)

# checking the created array:np_4
print(np_4)

[ 1.     2.125  3.25   4.375  5.5    6.625  7.75   8.875 10.   ]


## Another Numpy Functions

* ```np.power()```: Calculating powers of the array elements.
* ```np.absolute()```: Converting all the elements in the absolute form.
* ```np.sin()```or ```np.cos()```: Takes the sine/cosine of elements present in the array.
* ```np.log()```: Takes the log of the elements in the array.

In [111]:
np_1

array([1, 2, 3, 4])

In [114]:
np.power(np_1, 2)        # same as (np_1 ** 2)

array([ 1,  4,  9, 16], dtype=int32)

In [116]:
# getting the absolute value of the elements

np.absolute(np.array([-1, 2, -2, -0.99, 5]))

array([1.  , 2.  , 2.  , 0.99, 5.  ])

In [119]:
print(np_1)

print("np.sin(np_1):", np.sin(np_1))
print("np.sin(np_2):", np.cos(np_1))

[1 2 3 4]
np.sin(np_1): [ 0.84147098  0.90929743  0.14112001 -0.7568025 ]
np.sin(np_2): [ 0.54030231 -0.41614684 -0.9899925  -0.65364362]


In [120]:
np.log(np_1)

array([0.        , 0.69314718, 1.09861229, 1.38629436])

### Empty array

In [132]:
# Creating empty arrays

x = np.arange(1, 6)
y = np.empty(5)

print(x)
print(y)

[1 2 3 4 5]
[10. 20. 30. 40. 50.]


In [133]:
# get sum of whole array like fibonnacci

np.add.reduce(x)

15

array([ 1,  3,  6, 10, 15], dtype=int32)

### Initializing different types of arrays
* ```np.full()```: Create a constant array of any number ‘n’.
* ```np.char.tile()```: Create a new array, by repeating an existing array, for a fixed number of times.
* ```np.eye()```: Create an identity matrix of any dimension.
* ```np.random.random()```: Create an array of random numbers between 0 and 1.
* ```np.random.randint()```: Create a random array of integers within a particular range.

In [141]:
# Create array with any other numbers

print(np.full((2, 3), 10))
print(np.full(5, 2))
print(np.full([4, 3], 7))

[[10 10 10]
 [10 10 10]]
[2 2 2 2 2]
[[7 7 7]
 [7 7 7]
 [7 7 7]
 [7 7 7]]


In [144]:
np.char.title(np.array(['python', 'is', 'best', 'for', 'Data', 'Science']))

array(['Python', 'Is', 'Best', 'For', 'Data', 'Science'], dtype='<U7')

In [149]:
# Given an array, np.tile() creates a new array by repeating the given array for any number of times that you want

print("1. ", np.tile([1, 2, 3], 5))

# You can also create multidimensional arrays using np.tile()
print("2. ", np.tile(arr, (3,2)))

1.  [1 2 3 1 2 3 1 2 3 1 2 3 1 2 3]
2.  [[1 2 3 1 2 3]
 [1 2 3 1 2 3]
 [1 2 3 1 2 3]]


In [150]:
# np.eye() - The eye tool returns a 2-D array with  1’s as the diagonal and  0’s elsewhere. 
# The diagonal can be main, upper, or lower depending on the optional parameter k. 
# A positive k is for the upper diagonal, a negative k is for the lower, and a  0 k (default) is for the main diagonal.

print(np.eye(3))
print()
print(np.eye(3, k=1))
print()
print(np.eye(3, k=-1))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

[[0. 1. 0.]
 [0. 0. 1.]
 [0. 0. 0.]]

[[0. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]]


In [169]:
# returns an array of specified shape and fills it with random floats in the half-open interval [0.0, 1.0).
# returns values between 0 to 1 in floats

np.random.random((3, 2))

array([[0.50237478, 0.4935794 ],
       [0.3364078 , 0.35202125],
       [0.40471859, 0.46235478]])

In [171]:
# returns an array of specified shape and fills it with random integers from low (inclusive) to high (exclusive), 
# i.e. in the interval [low, high).

np.random.randint(low=1, high=10, size=(3, 3))

array([[5, 8, 4],
       [6, 6, 5],
       [8, 3, 6]])

In [17]:
# Create array with any other numbers with existing array size

np.full_like(a, 4)

array([[4, 4, 4, 4],
       [4, 4, 4, 4],
       [4, 4, 4, 4]])

In [137]:
# Create array with random decimal numbers

np.random.rand(4, 2)

array([[0.14487615, 0.17764428],
       [0.36891683, 0.80238371],
       [0.4511946 , 0.07655585],
       [0.6175487 , 0.28152226]])

In [20]:
# Identity matrix
np.identity(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [21]:
# Repeat an array

arr = np.array([[1, 2, 3]])
print(np.repeat(arr, 3, axis=0))
print(np.repeat(arr, 3, axis=1))

[[1 2 3]
 [1 2 3]
 [1 2 3]]
[[1 1 1 2 2 2 3 3 3]]


# Manipulating Arrays

Here are two techniques that can be used to change or manipulayte the structure of arrays:
*  Stacking
*  Reshaping

### Stacking

``` 
1. hstack(): horizontal stack, dimension of rows of two numpy array must be same
2. vstack(): vertical stack, dimension of columns of two numpy array must be same
```

In [172]:
# Creating two 1-D arrays with 5 elements using arange
arr_1 = np.arange(5)
arr_2 = np.arange(5, 10)

print(arr_1)
print(arr_2)

[0 1 2 3 4]
[5 6 7 8 9]


In [173]:
# Horizonal stacking - Appending the elements in the same row

np.hstack((arr_1, arr_2))

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [174]:
# Vertical stacking - Increasing the number of row

arr_3 = np.vstack((arr_1, arr_2))
arr_3

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

### Reshaping

In [175]:
# Creating an array of first ten natural numbers
arr_4 = np.arange(1, 11)

arr_4

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [177]:
# Reshaping the 1D array into a 2D array with 2 elements in five rows

arr_5 = arr_4.reshape(5, 2)
arr_5

array([[ 1,  2],
       [ 3,  4],
       [ 5,  6],
       [ 7,  8],
       [ 9, 10]])

In [178]:
# Creating a 2D array with 4 elements in five rows (elements from 20 to 39)
arr_6 = np.arange(20, 40)
print(arr_6)

arr_6 = arr_6.reshape(5, 4)
print(arr_6)

[20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39]
[[20 21 22 23]
 [24 25 26 27]
 [28 29 30 31]
 [32 33 34 35]
 [36 37 38 39]]


In [179]:
arr_6.reshape(10, 2)

array([[20, 21],
       [22, 23],
       [24, 25],
       [26, 27],
       [28, 29],
       [30, 31],
       [32, 33],
       [34, 35],
       [36, 37],
       [38, 39]])

In [190]:
# Stacking arr_5 and arr_6
arr_5 = np.arange(0, 20).reshape(5, 4)
print(arr_5)
print()

print(np.vstack((arr_5, arr_6)))

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]
 [16 17 18 19]]

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]
 [16 17 18 19]
 [20 21 22 23]
 [24 25 26 27]
 [28 29 30 31]
 [32 33 34 35]
 [36 37 38 39]]


In [192]:
# Index of the 100th Element
# Consider an (11,12) shape array. What is the index (x,y) of the 100th element? 
# Note: For counting the elements, go row-wise. For example, in the array:

print(np.arange(1, 133).reshape(11, 12))
print(np.unravel_index(99, (11,12)))

[[  1   2   3   4   5   6   7   8   9  10  11  12]
 [ 13  14  15  16  17  18  19  20  21  22  23  24]
 [ 25  26  27  28  29  30  31  32  33  34  35  36]
 [ 37  38  39  40  41  42  43  44  45  46  47  48]
 [ 49  50  51  52  53  54  55  56  57  58  59  60]
 [ 61  62  63  64  65  66  67  68  69  70  71  72]
 [ 73  74  75  76  77  78  79  80  81  82  83  84]
 [ 85  86  87  88  89  90  91  92  93  94  95  96]
 [ 97  98  99 100 101 102 103 104 105 106 107 108]
 [109 110 111 112 113 114 115 116 117 118 119 120]
 [121 122 123 124 125 126 127 128 129 130 131 132]]
(8, 3)


In [193]:
p = np.array([[1, 5],
 [3, 7],
 [4, 9]])

print(np.reshape(p, -1))
print(np.reshape(p, (1, -1)))
print(np.reshape(p, (1, -1)).ndim)

[1 5 3 7 4 9]
[[1 5 3 7 4 9]]
2


In [195]:
array_1 = np.array([[1, 2, 3, 4, 5],
 [6, 7, 8, 9, 10],
 [11, 12, 13, 14, 15],
 [16, 17, 18, 19, 20]])

print(array_1)
print()
print(array_1[array_1%2 != 0])

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]]

[ 1  3  5  7  9 11 13 15 17 19]


### Copy array

In [25]:
# Bad way of copying
a = np.array([1, 2, 3, 4])
b = a
print("a:", a)
print("b:", b)

b[0] = 100
print("a:", a)
print("b:", b)

a: [1 2 3 4]
b: [1 2 3 4]
a: [100   2   3   4]
b: [100   2   3   4]


In [27]:
# Bad way of copying
a = np.array([1, 2, 3, 4])
b = a.copy()
print("a:", a)
print("b:", b)

b[0] = 100
print("a:", a)
print("b:", b)

a: [1 2 3 4]
b: [1 2 3 4]
a: [1 2 3 4]
b: [100   2   3   4]


### Numpy Mathematics

In [28]:
a = np.array([1, 2, 3, 4])
a

array([1, 2, 3, 4])

In [29]:
a + 2

array([3, 4, 5, 6])

In [30]:
a - 2

array([-1,  0,  1,  2])

In [31]:
a * 2

array([2, 4, 6, 8])

In [32]:
a / 2

array([0.5, 1. , 1.5, 2. ])

In [36]:
a ** 2

array([ 1,  4,  9, 16], dtype=int32)

In [35]:
b = np.array([5, 6, 7, 8])
a + b

array([ 6,  8, 10, 12])

### Numpy Statistics

In [42]:
a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
a

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [43]:
a.max()

9

In [44]:
a.min()

1

In [47]:
# column
np.max(a, axis=1)

array([3, 6, 9])

In [48]:
# row
np.max(a, axis=0)

array([7, 8, 9])

In [49]:
# Sum, column wise
np.sum(a, axis=0)

array([12, 15, 18])

In [51]:
# Sum, row wise
np.sum(a, axis=1)

array([ 6, 15, 24])

### Reshaping array

In [54]:
a = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])
a

array([[1, 2, 3, 4],
       [5, 6, 7, 8]])

In [62]:
a.shape

(2, 4)

In [66]:
a.reshape(4, 2)

array([[1, 2],
       [3, 4],
       [5, 6],
       [7, 8]])

<hr />

## Practice Example 1

In [67]:
# player information is provided in the lists - players, skills
# players: list of tuples where 1st element is height in inches and the 2nd element is weight in lbs
# skills: the skill of the player in the sport cricket

players = [(74, 180), (74, 215), (72, 210), (72, 210), (73, 188), (69, 176), (69, 209), (71, 200), (76, 231), (71, 180), (73, 188), (73, 180), (74, 185), (74, 160), (69, 180), (70, 185), (73, 189), (75, 185), (78, 219), (79, 230), (76, 205), (74, 230), (76, 195), (72, 180), (71, 192), (75, 225), (77, 203), (74, 195), (73, 182), (74, 188), (78, 200), (73, 180), (75, 200), (73, 200), (75, 245), (75, 240), (74, 215), (69, 185), (71, 175), (74, 199), (73, 200), (73, 215), (76, 200), (74, 205), (74, 206), (70, 186), (72, 188), (77, 220), (74, 210), (70, 195), (73, 200), (75, 200), (76, 212), (76, 224), (78, 210), (74, 205), (74, 220), (76, 195), (77, 200), (81, 260), (78, 228), (75, 270), (77, 200), (75, 210), (76, 190), (74, 220), (72, 180), (72, 205), (75, 210), (73, 220), (73, 211), (73, 200), (70, 180), (70, 190), (70, 170), (76, 230), (68, 155), (71, 185), (72, 185), (75, 200), (75, 225), (75, 225), (75, 220), (68, 160), (74, 205), (78, 235), (71, 250), (73, 210), (76, 190), (74, 160), (74, 200), (79, 205), (75, 222), (73, 195), (76, 205), (74, 220), (74, 220), (73, 170), (72, 185), (74, 195), (73, 220), (74, 230), (72, 180), (73, 220), (69, 180), (72, 180), (73, 170), (75, 210), (75, 215), (73, 200), (72, 213), (72, 180), (76, 192), (74, 235), (72, 185), (77, 235), (74, 210), (77, 222), (75, 210), (76, 230), (80, 220), (74, 180), (74, 190), (75, 200), (78, 210), (73, 194), (73, 180), (74, 190), (75, 240), (76, 200), (71, 198), (73, 200), (74, 195), (76, 210), (76, 220), (74, 190), (73, 210), (74, 225), (70, 180), (72, 185), (73, 170), (73, 185), (73, 185), (73, 180), (71, 178), (74, 175), (74, 200), (72, 204), (74, 211), (71, 190), (74, 210), (73, 190), (75, 190), (75, 185), (79, 290), (73, 175), (75, 185), (76, 200), (74, 220), (76, 170), (78, 220), (74, 190), (76, 220), (72, 205), (74, 200), (76, 250), (74, 225), (75, 215), (78, 210), (75, 215), (72, 195), (74, 200), (72, 194), (74, 220), (70, 180), (71, 180), (70, 170), (75, 195), (71, 180), (71, 170), (73, 206), (72, 205), (71, 200), (73, 225), (72, 201), (75, 225), (74, 233), (74, 180), (75, 225), (73, 180), (77, 220), (73, 180), (76, 237), (75, 215), (74, 190), (76, 235), (75, 190), (73, 180), (71, 165), (76, 195), (75, 200), (72, 190), (71, 190), (77, 185), (73, 185), (74, 205), (71, 190), (72, 205), (74, 206), (75, 220), (73, 208), (72, 170), (75, 195), (75, 210), (74, 190), (72, 211), (74, 230), (71, 170), (70, 185), (74, 185), (77, 241), (77, 225), (75, 210), (75, 175), (78, 230), (75, 200), (76, 215), (73, 198), (75, 226), (75, 278), (79, 215), (77, 230), (76, 240), (71, 184), (75, 219), (74, 170), (69, 218), (71, 190), (76, 225), (72, 220), (72, 176), (70, 190), (72, 197), (73, 204), (71, 167), (72, 180), (71, 195), (73, 220), (72, 215), (73, 185), (74, 190), (74, 205), (72, 205), (75, 200), (74, 210), (74, 215), (77, 200), (75, 205), (73, 211), (72, 190), (71, 208), (74, 200), (77, 210), (75, 232), (75, 230), (75, 210), (78, 220), (78, 210), (74, 202), (76, 212), (78, 225), (76, 170), (70, 190), (72, 200), (80, 237), (74, 220), (74, 170), (71, 193), (70, 190), (72, 150), (71, 220), (74, 200), (71, 190), (72, 185), (71, 185), (74, 200), (69, 172), (76, 220), (75, 225), (75, 190), (76, 195), (73, 219), (76, 190), (73, 197), (77, 200), (73, 195), (72, 210), (72, 177), (77, 220), (77, 235), (71, 180), (74, 195), (74, 195), (73, 190), (78, 230), (75, 190), (73, 200), (70, 190), (74, 190), (72, 200), (73, 200), (73, 184), (75, 200), (75, 180), (74, 219), (76, 187), (73, 200), (74, 220), (75, 205), (75, 190), (72, 170), (73, 160), (73, 215), (72, 175), (74, 205), (78, 200), (76, 214), (73, 200), (74, 190), (75, 180), (70, 205), (75, 220), (71, 190), (72, 215), (78, 235), (75, 191), (73, 200), (73, 181), (71, 200), (75, 210), (77, 240), (72, 185), (69, 165), (73, 190), (74, 185), (72, 175), (70, 155), (75, 210), (70, 170), (72, 175), (72, 220), (74, 210), (73, 205), (74, 200), (76, 205), (75, 195), (80, 240), (72, 150), (75, 200), (73, 215), (74, 202), (74, 200), (73, 190), (75, 205), (75, 190), (71, 160), (73, 215), (75, 185), (74, 200), (74, 190), (72, 210), (74, 185), (74, 220), (74, 190), (73, 202), (76, 205), (75, 220), (72, 175), (73, 160), (73, 190), (73, 200), (72, 229), (72, 206), (72, 220), (72, 180), (71, 195), (75, 175), (75, 188), (74, 230), (73, 190), (75, 200), (79, 190), (74, 219), (76, 235), (73, 180), (74, 180), (74, 180), (72, 200), (74, 234), (74, 185), (75, 220), (78, 223), (74, 200), (74, 210), (74, 200), (77, 210), (70, 190), (73, 177), (74, 227), (73, 180), (71, 195), (75, 199), (71, 175), (72, 185), (77, 240), (74, 210), (70, 180), (77, 194), (73, 225), (72, 180), (76, 205), (71, 193), (76, 230), (78, 230), (75, 220), (73, 200), (78, 249), (74, 190), (79, 208), (75, 245), (76, 250), (72, 160), (75, 192), (75, 220), (70, 170), (72, 197), (70, 155), (74, 190), (71, 200), (76, 220), (73, 210), (76, 228), (71, 190), (69, 160), (72, 184), (72, 180), (69, 180), (73, 200), (69, 176), (73, 160), (74, 222), (74, 211), (72, 195), (71, 200), (72, 175), (72, 206), (76, 240), (76, 185), (76, 260), (74, 185), (76, 221), (75, 205), (71, 200), (72, 170), (71, 201), (73, 205), (75, 185), (76, 205), (75, 245), (71, 220), (75, 210), (74, 220), (72, 185), (73, 175), (73, 170), (73, 180), (73, 200), (76, 210), (72, 175), (76, 220), (73, 206), (73, 180), (73, 210), (75, 195), (75, 200), (77, 200), (73, 164), (72, 180), (75, 220), (70, 195), (74, 205), (72, 170), (80, 240), (71, 210), (71, 195), (74, 200), (74, 205), (73, 192), (75, 190), (76, 170), (73, 240), (77, 200), (72, 205), (73, 175), (77, 250), (76, 220), (71, 224), (75, 210), (73, 195), (74, 180), (77, 245), (71, 175), (72, 180), (73, 215), (69, 175), (73, 180), (70, 195), (74, 230), (76, 230), (73, 205), (73, 215), (75, 195), (73, 180), (79, 205), (74, 180), (73, 190), (74, 180), (77, 190), (75, 190), (74, 220), (73, 210), (77, 255), (73, 190), (77, 230), (74, 200), (74, 205), (73, 210), (77, 225), (74, 215), (77, 220), (75, 205), (77, 200), (75, 220), (71, 197), (74, 225), (70, 187), (79, 245), (72, 185), (72, 185), (70, 175), (74, 200), (74, 180), (72, 188), (73, 225), (72, 200), (74, 210), (74, 245), (76, 213), (82, 231), (74, 165), (74, 228), (70, 210), (73, 250), (73, 191), (74, 190), (77, 200), (72, 215), (76, 254), (73, 232), (73, 180), (72, 215), (74, 220), (74, 180), (71, 200), (72, 170), (75, 195), (74, 210), (74, 200), (77, 220), (70, 165), (71, 180), (73, 200), (76, 200), (71, 170), (75, 224), (74, 220), (72, 180), (76, 198), (79, 240), (76, 239), (73, 185), (76, 210), (78, 220), (75, 200), (76, 195), (72, 220), (72, 230), (73, 170), (73, 220), (75, 230), (71, 165), (76, 205), (70, 192), (75, 210), (74, 205), (75, 200), (73, 210), (71, 185), (71, 195), (72, 202), (73, 205), (73, 195), (72, 180), (69, 200), (73, 185), (78, 240), (71, 185), (73, 220), (75, 205), (76, 205), (70, 180), (74, 201), (77, 190), (75, 208), (79, 240), (72, 180), (77, 230), (73, 195), (75, 215), (75, 190), (75, 195), (73, 215), (73, 215), (76, 220), (77, 220), (75, 230), (70, 195), (71, 190), (71, 195), (75, 209), (74, 204), (69, 170), (70, 185), (75, 205), (72, 175), (75, 210), (73, 190), (72, 180), (72, 180), (72, 160), (76, 235), (75, 200), (74, 210), (69, 180), (73, 190), (72, 197), (72, 203), (75, 205), (77, 170), (76, 200), (80, 250), (77, 200), (76, 220), (79, 200), (71, 190), (75, 170), (73, 190), (76, 220), (77, 215), (73, 206), (76, 215), (70, 185), (75, 235), (73, 188), (75, 230), (70, 195), (69, 168), (71, 190), (72, 160), (72, 200), (73, 200), (70, 189), (70, 180), (73, 190), (76, 200), (75, 220), (72, 187), (73, 240), (79, 190), (71, 180), (72, 185), (74, 210), (74, 220), (74, 219), (72, 190), (76, 193), (76, 175), (72, 180), (72, 215), (71, 210), (72, 200), (72, 190), (70, 185), (77, 220), (74, 170), (72, 195), (76, 205), (71, 195), (76, 210), (71, 190), (73, 190), (70, 180), (73, 220), (73, 190), (72, 186), (71, 185), (71, 190), (71, 180), (72, 190), (72, 170), (74, 210), (74, 240), (74, 220), (71, 180), (72, 210), (75, 210), (72, 195), (71, 160), (72, 180), (72, 205), (72, 200), (72, 185), (74, 245), (74, 190), (77, 210), (75, 200), (73, 200), (75, 222), (73, 215), (76, 240), (72, 170), (77, 220), (75, 156), (72, 190), (71, 202), (71, 221), (75, 200), (72, 190), (73, 210), (73, 190), (71, 200), (70, 165), (75, 190), (71, 185), (76, 230), (73, 208), (68, 209), (71, 175), (72, 180), (74, 200), (77, 205), (72, 200), (76, 250), (78, 210), (81, 230), (72, 244), (73, 202), (76, 240), (72, 200), (72, 215), (74, 177), (76, 210), (73, 170), (76, 215), (75, 217), (70, 198), (71, 200), (74, 220), (72, 170), (73, 200), (76, 230), (76, 231), (73, 183), (71, 192), (68, 167), (71, 190), (71, 180), (74, 180), (77, 215), (69, 160), (72, 205), (76, 223), (75, 175), (76, 170), (75, 190), (76, 240), (72, 175), (74, 230), (76, 223), (74, 196), (72, 167), (75, 195), (78, 190), (77, 250), (70, 190), (72, 190), (79, 190), (74, 170), (71, 160), (68, 150), (77, 225), (75, 220), (71, 209), (72, 210), (70, 176), (72, 260), (72, 195), (73, 190), (72, 184), (74, 180), (72, 195), (72, 195), (75, 219), (72, 225), (73, 212), (74, 202), (72, 185), (78, 200), (75, 209), (72, 200), (74, 195), (75, 228), (75, 210), (76, 190), (74, 212), (74, 190), (73, 218), (74, 220), (71, 190), (74, 235), (75, 210), (76, 200), (74, 188), (76, 210), (76, 235), (73, 188), (75, 215), (75, 216), (74, 220), (68, 180), (72, 185), (75, 200), (71, 210), (70, 220), (72, 185), (73, 231), (72, 210), (75, 195), (74, 200), (70, 205), (76, 200), (71, 190), (82, 250), (72, 185), (73, 180), (74, 170), (71, 180), (75, 208), (77, 235), (72, 215), (74, 244), (72, 220), (73, 185), (78, 230), (77, 190), (73, 200), (73, 180), (73, 190), (73, 196), (73, 180), (76, 230), (75, 224), (70, 160), (73, 178), (72, 205), (73, 185), (75, 210), (74, 180), (73, 190), (73, 200), (76, 257), (73, 190), (75, 220), (70, 165), (77, 205), (72, 200), (77, 208), (74, 185), (75, 215), (75, 170), (75, 235), (75, 210), (72, 170), (74, 180), (71, 170), (76, 190), (71, 150), (75, 230), (76, 203), (83, 260), (75, 246), (74, 186), (76, 210), (72, 198), (72, 210), (75, 215), (75, 180), (72, 200), (77, 245), (73, 200), (72, 192), (70, 192), (74, 200), (72, 192), (74, 205), (72, 190), (71, 186), (70, 170), (71, 197), (76, 219), (74, 200), (76, 220), (74, 207), (74, 225), (74, 207), (75, 212), (75, 225), (71, 170), (71, 190), (74, 210), (77, 230), (71, 210), (74, 200), (75, 238), (77, 234), (76, 222), (74, 200), (76, 190), (72, 170), (71, 220), (72, 223), (75, 210), (73, 215), (68, 196), (72, 175), (69, 175), (73, 189), (73, 205), (75, 210), (70, 180), (70, 180), (74, 197), (75, 220), (74, 228), (74, 190), (73, 204), (74, 165), (75, 216), (77, 220), (73, 208), (74, 210), (76, 215), (74, 195), (75, 200), (73, 215), (76, 229), (78, 240), (75, 207), (73, 205), (77, 208), (74, 185), (72, 190), (74, 170), (72, 208), (71, 225), (73, 190), (75, 225), (73, 185), (67, 180), (67, 165), (76, 240), (74, 220), (73, 212), (70, 163), (75, 215), (70, 175), (72, 205), (77, 210), (79, 205), (78, 208), (74, 215), (75, 180), (75, 200), (78, 230), (76, 211), (75, 230), (69, 190), (75, 220), (72, 180), (75, 205), (73, 190), (74, 180), (75, 205), (75, 190), (73, 195)]
skills = np.array(['Keeper', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Bowler', 'Bowler', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Keeper', 'Keeper', 'Keeper', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Keeper', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Bowler', 'Bowler', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper', 'Batsman', 'Keeper', 'Bowler', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Bowler', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Batsman', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Bowler', 'Bowler', 'Bowler', 'Batsman', 'Keeper', 'Bowler', 'Bowler', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Batsman', 'Bowler', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper', 'Keeper', 'Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Batsman', 'Bowler', 'Keeper', 'Keeper', 'Batsman', 'Bowler', 'Bowler', 'Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Bowler', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Batsman', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Batsman', 'Batsman', 'Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Keeper', 'Bowler', 'Keeper', 'Bowler', 'Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Batsman', 'Keeper', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Batsman', 'Bowler', 'Keeper', 'Bowler', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Bowler', 'Bowler', 'Bowler', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Batsman', 'Bowler', 'Keeper', 'Bowler', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Batsman', 'Batsman', 'Keeper', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Bowler', 'Batsman', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Bowler', 'Bowler', 'Keeper', 'Bowler', 'Keeper', 'Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Bowler', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Batsman', 'Batsman', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Batsman', 'Batsman', 'Bowler', 'Batsman', 'Keeper', 'Bowler', 'Bowler', 'Keeper', 'Bowler', 'Bowler', 'Keeper', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Keeper-Batsman', 'Keeper-Batsman', 'Bowler', 'Bowler', 'Bowler', 'Batsman', 'Bowler', 'Keeper-Batsman', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper', 'Keeper-Batsman', 'Batsman', 'Bowler', 'Keeper-Batsman'])


In [68]:
# Create 2D array with list of tuples, e.g-players

np_players = np.array(players)
np_players

array([[ 74, 180],
       [ 74, 215],
       [ 72, 210],
       ...,
       [ 75, 205],
       [ 75, 190],
       [ 73, 195]])

In [70]:
type(np_players)

numpy.ndarray

In [73]:
# Get dimension (1D or 2D)

np_players.ndim

2

In [74]:
# Get number of rows and columns in array

np_players.shape

(1015, 2)

In [76]:
# Get data types of element in array

np_players.dtype

dtype('int32')

### Slicing array

In [78]:
# Get entire second row

np_players[1]

array([ 74, 215])

In [79]:
np_players[1][1]

215

In [81]:
# Get only first column value i.e height of players

np_players[:, 0]

array([74, 74, 72, ..., 75, 75, 73])

In [91]:
# Get only those players whose height is > 75

x = np_players[np_players[:, 0] > 75]
x

array([[ 76, 231],
       [ 78, 219],
       [ 79, 230],
       [ 76, 205],
       [ 76, 195],
       [ 77, 203],
       [ 78, 200],
       [ 76, 200],
       [ 77, 220],
       [ 76, 212],
       [ 76, 224],
       [ 78, 210],
       [ 76, 195],
       [ 77, 200],
       [ 81, 260],
       [ 78, 228],
       [ 77, 200],
       [ 76, 190],
       [ 76, 230],
       [ 78, 235],
       [ 76, 190],
       [ 79, 205],
       [ 76, 205],
       [ 76, 192],
       [ 77, 235],
       [ 77, 222],
       [ 76, 230],
       [ 80, 220],
       [ 78, 210],
       [ 76, 200],
       [ 76, 210],
       [ 76, 220],
       [ 79, 290],
       [ 76, 200],
       [ 76, 170],
       [ 78, 220],
       [ 76, 220],
       [ 76, 250],
       [ 78, 210],
       [ 77, 220],
       [ 76, 237],
       [ 76, 235],
       [ 76, 195],
       [ 77, 185],
       [ 77, 241],
       [ 77, 225],
       [ 78, 230],
       [ 76, 215],
       [ 79, 215],
       [ 77, 230],
       [ 76, 240],
       [ 76, 225],
       [ 77,

In [92]:
x.shape

(207, 2)

In [95]:
# Get those rows where the skill of the player is 'Batsman'

y = np_players[skills == "Batsman"]
y

array([[ 74, 215],
       [ 73, 188],
       [ 69, 209],
       [ 71, 200],
       [ 76, 231],
       [ 73, 188],
       [ 73, 180],
       [ 70, 185],
       [ 79, 230],
       [ 72, 180],
       [ 71, 192],
       [ 75, 225],
       [ 77, 203],
       [ 73, 182],
       [ 75, 245],
       [ 74, 215],
       [ 71, 175],
       [ 73, 200],
       [ 74, 205],
       [ 76, 195],
       [ 75, 210],
       [ 76, 190],
       [ 72, 180],
       [ 75, 210],
       [ 73, 220],
       [ 70, 180],
       [ 70, 190],
       [ 71, 185],
       [ 72, 185],
       [ 75, 200],
       [ 78, 235],
       [ 71, 250],
       [ 73, 210],
       [ 73, 195],
       [ 76, 205],
       [ 74, 220],
       [ 73, 170],
       [ 72, 185],
       [ 73, 220],
       [ 72, 180],
       [ 73, 220],
       [ 72, 180],
       [ 75, 210],
       [ 75, 215],
       [ 72, 213],
       [ 72, 180],
       [ 77, 235],
       [ 77, 222],
       [ 75, 210],
       [ 76, 230],
       [ 80, 220],
       [ 75, 200],
       [ 73,

In [96]:
y.shape

(323, 2)

In [196]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

## Practice Exercise 2

You are provided with 2 lists that contain the data of an ecommerce website. 
The first list contains the data for the number of items sold for a particular product and the second list contains the price of the product sold. As a part of this exercise, solve the questions that are provided below.

In [197]:
number = [8, 9, 9, 1, 6, 9, 5, 7, 3, 9, 7, 3, 4, 8, 3, 5, 8, 4, 8, 7, 5, 7, 3, 6, 1, 2, 7, 4, 7, 7, 8, 4, 3, 4, 2, 2, 2, 7, 3, 5, 6, 1, 1, 3, 2, 1, 1, 7, 7, 1, 4, 4, 5, 6, 1, 2, 7, 4, 5, 8, 1, 4, 8, 6, 2, 4, 3, 7, 3, 6, 2, 3, 3, 3, 2, 4, 6, 8, 9, 3, 9, 3, 1, 8, 6, 6, 3, 3, 9, 4, 6, 4, 9, 6, 7, 1, 2, 8, 7, 8, 1, 4]
price = [195, 225, 150, 150, 90, 60, 75, 255, 270, 225, 135, 195, 30, 15, 210, 105, 15, 30, 180, 60, 165, 60, 45, 225, 180, 90, 30, 210, 150, 15, 270, 60, 210, 180, 60, 225, 150, 150, 120, 195, 75, 240, 60, 45, 30, 180, 240, 285, 135, 165, 180, 240, 60, 105, 165, 240, 120, 45, 120, 165, 285, 225, 90, 105, 225, 45, 45, 45, 75, 180, 90, 240, 30, 30, 60, 135, 180, 15, 255, 180, 270, 135, 105, 135, 210, 180, 135, 195, 225, 75, 225, 15, 240, 60, 15, 180, 255, 90, 15, 150, 230, 150]

number = np.array(number)
price = np.array(price)

In [199]:
print(number)
print()
print(price)

[8 9 9 1 6 9 5 7 3 9 7 3 4 8 3 5 8 4 8 7 5 7 3 6 1 2 7 4 7 7 8 4 3 4 2 2 2
 7 3 5 6 1 1 3 2 1 1 7 7 1 4 4 5 6 1 2 7 4 5 8 1 4 8 6 2 4 3 7 3 6 2 3 3 3
 2 4 6 8 9 3 9 3 1 8 6 6 3 3 9 4 6 4 9 6 7 1 2 8 7 8 1 4]

[195 225 150 150  90  60  75 255 270 225 135 195  30  15 210 105  15  30
 180  60 165  60  45 225 180  90  30 210 150  15 270  60 210 180  60 225
 150 150 120 195  75 240  60  45  30 180 240 285 135 165 180 240  60 105
 165 240 120  45 120 165 285 225  90 105 225  45  45  45  75 180  90 240
  30  30  60 135 180  15 255 180 270 135 105 135 210 180 135 195 225  75
 225  15 240  60  15 180 255  90  15 150 230 150]


#### Questions

#### 1. How many different products are sold by the company in total?

-  99
-  100
-  101
-  102


In [201]:
number.size

102

#### 2. How many items were sold in total?

-  460
-  490
-  500
-  520


In [203]:
number.sum()

490

#### 3. What is the average price of the products sold by the ecommerce company?

-  139
-  151
-  142
-  128

In [205]:
price.mean()

139.01960784313727


#### 4. What is the price of the costliest item sold?

-  225
-  310
-  280
-  285

In [206]:
price.max()

285

#### 5. What is the total revenue of the company? [Revenue = Price\*Quantity]

-  67100
-  53900
-  45300
-  71200

In [207]:
(price * number).sum()

67100